AddressBook.java

```diff
@@ -0,0 +1,1317 @@
+     * @return person's name
+     */
+    private static String getNameFromPerson(HashMap<personProperty, String> person) {
+         return person.get(personProperty.NAME);
+    	/*return person[PERSON_DATA_INDEX_NAME];*/
+    }
+
+    /**
+     * @param person whose phone number you want
+     * @return person's phone number
+     */
+    private static String getPhoneFromPerson(HashMap<personProperty, String> person) {
+    	 return person.get(personProperty.PHONE);
+        /*return person[PERSON_DATA_INDEX_PHONE];*/
+    }
+
+    /**
+     * @param person whose email you want
+     * @return person's email
+     */
+    private static String getEmailFromPerson(HashMap<personProperty, String> person) {
+    	return person.get(personProperty.EMAIL);
+    	/* return person[PERSON_DATA_INDEX_EMAIL];*/
+    }
+
+    /**
+     * Create a person for use in the internal data.
+     *
+     * @param name of person
+     * @param phone without data prefix
+     * @param email without data prefix
+     * @return constructed person
+     */
+    private static String[] makePersonFromData(String name, String phone, String email) {
+        final String[] person = new String[PERSON_DATA_COUNT];
+        person[PERSON_DATA_INDEX_NAME] = name.toLowerCase();
+        person[PERSON_DATA_INDEX_PHONE] = phone;
+        person[PERSON_DATA_INDEX_EMAIL] = email;
+        return person;
+    }
+
+    /**
+     * Encodes a person into a decodable and readable string representation.
+     *
+     * @param person to be encoded
+     * @return encoded string
+     */
+    private static String encodePersonToString(HashMap<personProperty, String> person) {
+        return String.format(PERSON_STRING_REPRESENTATION,
+                getNameFromPerson(person), getPhoneFromPerson(person), getEmailFromPerson(person));
+    }
+
+    /**
+     * Encodes list of persons into list of decodable and readable string representations.
+     *
+     * @param persons to be encoded
+     * @return encoded strings
+     */
+    private static ArrayList<String> encodePersonsToStrings(ArrayList<HashMap<personProperty, String>> persons) {
+        final ArrayList<String> encoded = new ArrayList<String>();
+        for (HashMap<personProperty, String> person : persons) {
+            encoded.add(encodePersonToString(person));
+        }
+        return encoded;
+    }
+
+    /*
+     * ==============NOTE TO STUDENTS======================================
+     * Note the use of Java's new 'Optional' feature to indicate that
+     * the return value may not always be present.
+     * ====================================================================
+     */
+    /**
+     * Decodes a person from it's supposed string representation.
+     *
+     * @param encoded string to be decoded
+     * @return if cannot decode: empty Optional
+     *         else: Optional containing decoded person
+     */
+    private static Optional<String[]> decodePersonFromString(String encoded) {
```

Comment:
better if this returns a HashMap instead of String[] since you're now representing people using HashMap to represent persons